buffer.c

```diff
@@ -224,55 +139,59 @@ PHP_METHOD(ArrayBuffer, serialize)
 	smart_str buf = {0};
 	php_serialize_data_t var_hash;
 	zval zv;
-	zval *zv_ptr = &zv;
 
 	if (zend_parse_parameters_none() == FAILURE) {
 		return;
 	}
 
-	intern = zend_object_store_get_object(getThis() TSRMLS_CC);
+	intern = Z_BUFFER_OBJ_P(getThis());
 	if (!intern->buffer) {
 		return;
 	}
 
 	PHP_VAR_SERIALIZE_INIT(var_hash);
 
-	INIT_PZVAL(zv_ptr);
-
 	/* Serialize buffer as string */
-	ZVAL_STRINGL(zv_ptr, (char *) intern->buffer, (int) intern->length, 0);
-	php_var_serialize(&buf, &zv_ptr, &var_hash TSRMLS_CC);
+        zend_string * zstr = zend_string_init((char *) intern->buffer, (int) intern->length, 0);
+	ZVAL_STR(&zv, zstr);
+	php_var_serialize(&buf, &zv, &var_hash);
+	zend_string_release(zstr);
 
 	/* Serialize properties as array */
-	Z_ARRVAL_P(zv_ptr) = zend_std_get_properties(getThis() TSRMLS_CC);
-	Z_TYPE_P(zv_ptr) = IS_ARRAY;
-	php_var_serialize(&buf, &zv_ptr, &var_hash TSRMLS_CC);
+	ZVAL_ARR(&zv, zend_std_get_properties(getThis()));
+	php_var_serialize(&buf, &zv, &var_hash);
 
 	PHP_VAR_SERIALIZE_DESTROY(var_hash);
 
-	if (buf.c) {
-		RETURN_STRINGL(buf.c, buf.len, 0);
-	}
+        if (buf.s) {
+                RETURN_NEW_STR(buf.s);
+        }
+
+        RETURN_NULL();
 }
 
 PHP_METHOD(ArrayBuffer, unserialize)
 {
 	buffer_object *intern;
 	char *str;
-	int str_len;
+	size_t str_len;
 	php_unserialize_data_t var_hash;
 	const unsigned char *p, *max;
-	zval zv, *zv_ptr = &zv;
+	zval * zbuf, * ztable;
+        zend_string * zstr;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s", &str, &str_len) == FAILURE) {
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "s", &str, &str_len) == FAILURE) {
+		return;
+	}
+	if (str_len == 0) {
```

Comment:
Hm, why do we need this check?

Reply:
Looking at `zend_API.h`, I think it might be because this is asking for a pointer to a char with `"s"` the code will return a NULL for `str` when `str_len==0`:

```c
static zend_always_inline bool zend_parse_arg_string(zval *arg, char **dest, size_t *dest_len, bool check_null, uint32_t arg_num)
{
        zend_string *str;

        if (!zend_parse_arg_str(arg, &str, check_null, arg_num)) {
                return 0;
        }
        if (check_null && UNEXPECTED(!str)) {
                *dest = NULL;
                *dest_len = 0;
        } else {
```

That NULL pointer might choke the call

```c
	if (!php_var_unserialize(zbuf, &p, max, &var_hash)) {
```

(Not familiar how `php_var_unserialize` will handle `p` pointing to variable pointing to NULL.

Reply:
I believe str can only be null if `s!` is used (i.e a nullable parameter). With `s` it should be always non-null. For str_len == 0, it would just be an empty string.

Reply:
Oh, yes, I missed the `check_null` part.

Removed the check.